apis/iam/v1beta1/diff_test.go

```diff
@@ -0,0 +1,856 @@
+					{
+						Role:    "roles/viewer",
+						Members: []Member{"user:viewer@example.com"},
+					},
+					{
+						Role:    "roles/editor",
+						Members: []Member{"user:editor@example.com"},
+					},
+				},
+				AuditConfigs: []IAMPolicyAuditConfig{
+					{
+						Service: "allServices",
+						AuditLogConfigs: []AuditLogConfig{
+							{
+								LogType: "DATA_READ",
+							},
+						},
+					},
+				},
+			},
+			actual: &IAMPolicySpec{
+				ResourceReference: ResourceReference{
+					Kind:      "Folder",
+					Namespace: "other-namespace",
+					Name:      "other-project",
+					External:  "projects/other-project",
+				},
+				Bindings: []IAMPolicyBinding{
+					{
+						Role:    "roles/viewer",
+						Members: []Member{"user:viewer-actual@example.com"},
+					},
+				},
+				AuditConfigs: []IAMPolicyAuditConfig{
+					{
+						Service: "compute.googleapis.com",
+						AuditLogConfigs: []AuditLogConfig{
+							{
+								LogType: "DATA_WRITE",
+							},
+						},
+					},
+				},
+			},
+			want: &structuredreporting.Diff{
+				Fields: []structuredreporting.DiffField{
+					{
+						ID:  "spec.resourceRef.kind",
+						Old: "Project",
+						New: "Folder",
+					},
+					{
+						ID:  "spec.resourceRef.namespace",
+						Old: "test-namespace",
+						New: "other-namespace",
+					},
+					{
+						ID:  "spec.resourceRef.name",
+						Old: "test-project",
+						New: "other-project",
+					},
+					{
+						ID:  "spec.resourceRef.external",
+						Old: "projects/test-project",
+						New: "projects/other-project",
+					},
+					{
+						ID:  "spec.bindings[role=roles/editor]",
+						Old: "present in desired spec",
+						New: "absent in actual spec",
+					},
+					{
+						ID: "spec.bindings[role=roles/viewer]",
+						Old: IAMPolicyBinding{
+							Role:    "roles/viewer",
+							Members: []Member{"user:viewer@example.com"},
+						},
+						New: IAMPolicyBinding{
+							Role:    "roles/viewer",
+							Members: []Member{"user:viewer-actual@example.com"},
+						},
+					},
+					{
+						ID:  "spec.auditConfigs[service=allServices]",
+						Old: "present in desired spec",
+						New: "absent in actual spec",
+					},
+					{
+						ID:  "spec.auditConfigs[service=compute.googleapis.com]",
+						Old: "absent in desired spec",
+						New: "present in actual spec",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Create a deep copy of the original desired spec for later comparison
+			originalDesired := tc.desired.DeepCopy()
+
+			got := IAMPolicySpecDiffers(tc.desired, tc.actual)
+
+			if diff := cmp.Diff(originalDesired, tc.desired); diff != "" {
+				t.Errorf("IAMPolicySpecDiffers() mutated the desired spec (-original +mutated):\n%v", diff)
+			}
+
+			if diff := cmp.Diff(sortDiffFields(tc.want), sortDiffFields(got)); diff != "" {
+				t.Errorf("IAMPolicySpecDiffers() got diff (-want +got):\n%v", diff)
+			}
+		})
```

Comment:
The test correctly checks that the `desired` spec is not mutated. For completeness and to guard against future regressions, it would be a good practice to also assert that the `actual` spec is not mutated by the `IAMPolicySpecDiffers` function.